torch/_inductor/codegen/cpp_wrapper_cpu.py

```diff
@@ -1057,16 +1057,21 @@ def generate_c_shim_extern_kernel_call(self, kernel, args):
                 f"AOTI_TORCH_ERROR_CODE_CHECK({shim_fn}({', '.join(args)}));"
             )
 
-    def generate_c_shim_extern_kernel_alloc(self, extern_kernel, args):
+    def generate_c_shim_extern_kernel_alloc(
+        self, extern_kernel: ir.ExternKernelAlloc, args: list[str]
+    ) -> None:
         # registered output buffer name
         name = extern_kernel.name
         output_handle_name = f"{name}_handle"
-        self.writeline(f"AtenTensorHandle {output_handle_name};")
-        output_arg = f"&{output_handle_name}"
-        self.generate_c_shim_extern_kernel_call(
-            extern_kernel.get_kernel_name(), args + [output_arg]
-        )
-        self.writeline(f"RAIIAtenTensorHandle {name}({output_handle_name});")
+        # In-place returns are automatically excised from the torchgen-ed shim.  There
+        # may be a better way to determine whether this op is in-place, but checking for
+        # input aliasing pragmatically seems to work.
+        if not (aliased_inputs := extern_kernel.get_inputs_that_alias_output()):
```

Comment:
I am not sure if this is the best way to check in-place op. @eellison , any recommendation here?

Reply:
By inplace, is this meant to be only ops that return the input, or aliasing ops ? `inputs_that_alias_output` will also refer to aliasing but I think here you only want ops that exactly return the input right ?

Reply:
@eellison It's ill-defined, at best. I've dug into `gen_aoti_c_shim.py`; it appears to omit all output arguments from the C-shim API if an `inplace` boolean on the schema is set. Best I can tell, that boolean is set in `torchgen.model.BaseOperatorName.parse()` if the method name fits either of the patterns `__i*__` (iff * is an enumerated arithmetic operator name) or `*_` (unconditionally, it seems). I can simply re-implement that check, but tbh it seems fragile at best?

Reply:
If it's dependent on the implementation details of `gen_aoti_c_shim` then @desertfire is a better reviewer here.

Other misc that could help:
- `set_` has tag `inplace_view` which means we mutated the input tensor metadata and return it. so ops for this are guaranteed to return the input.
- generally, aten operators that have are suffixed with `_` and mark mutation should return the input. but, these should be functionalized away. i believe it's only `set_` that appears of this category.
- you could also look V.graph.current_node and check that the output val is the same val as the input.


cc @bdhirsh

Reply:
If `inplace` is not available in the op schema, I think checking the `inplace_view` tag is a reasonable choice here.

Reply:
@desertfire @eellison Rewrote using the `inplace_view` tag.